Decoupling data table manager components (#2807)
commercetools/ui-kit: packages/components/data-table-manager/README.md

- The `disableDisplaySettings` enables / disables the layout settings panel, allowing the user to select wrapping text and density display options.
- The `disableColumnManager` enables / disables the column manager panel, allowing the user to select which columns are visible.

> - To Detach the `DatatableManager` settings dropdown from the Datatable and position it anywhere else, you would need to import a `DataTableManagerProvider` and wrap both components with the provider.

Reviewer: I believe we would need a dedicated section to explain this behaviour.

Author: Updated here: https://github.com/commercetools/ui-kit/pull/2807/commits/bd8447871f0d99ac366b5adc9d865055c9a4506e

Thank you

Both panels delegate the handling of the settings change on the parent through function properties, allowing the settings to be persisted or just used as state props.

> - The `children` prop can only be used when the datatable manager is used in it's coupled or undetached form, and not when it is detached.

Reviewer: This is my proposal for updating the README:
```
# DataTableManager

## Description

This component enhances the `<DataTable>` component and additionally provides a UI and state management to handle configuration of the table such as column manager.

- The `disableDisplaySettings` enables / disables the layout settings panel, allowing the user to select wrapping text and density display options.
- The `disableColumnManager` enables / disables the column manager panel, allowing the user to select which columns are visible.

Both panels delegate the handling of the settings change on the parent through function properties, allowing the settings to be persisted or just used as state props.

This component will render a triggering element (icon button) above the `<DataTable>` (top-right corner) which users can click to select the panel to open.

## Installation

```
yarn add @commercetools-uikit/data-table-manager
```

```
npm --save install @commercetools-uikit/data-table-manager
```

Additionally install the peer dependencies (if not present)

```
yarn add react react-dom react-intl
```

```
npm --save install react react-dom react-intl
```

## Usage

There are two ways this component can be used:

### Attached

The basic usage is when it just wraps the DataTable component.

```jsx
import DataTableManager from '@commercetools-uikit/data-table-manager';
import DataTable from '@commercetools-uikit/data-table';

const rows = [
  { id: 'parasite', title: 'Parasite', country: 'South Korea' },
  { id: 'portrait', title: 'Portrait of a Lady on Fire', country: 'France' },
  { id: 'wat', title: 'Woman at War', country: 'Iceland' },
];

const columns = [
  { key: 'title', label: 'Title' },
  { key: 'country', label: 'Country' },
];

const Example = () => (
  <DataTableManager columns={columns}>
    <DataTable rows={rows} />
  </DataTableManager>
);

export default Example;
```

### Dettached

As mentioned earlier, the default behaviour is for the triggering element to be placed above the `<DataTable>` component (top-right corner), but there can be use cases where it would be needed to place that triggering element in a different place in the page. This is also possible but requires the usage of one more component (`DataTableManagerProvider`) in order to share the manager state between the manager panels and the `DataTable` component.

```jsx
import DataTableManager, {
  DataTableManagerProvider,
} from '@commercetools-uikit/data-table-manager';
import DataTable from '@commercetools-uikit/data-table';
import SearchTextInput from '@commercetools-uikit/search-text-input';

const rows = [
  { id: 'parasite', title: 'Parasite', country: 'South Korea' },
  { id: 'portrait', title: 'Portrait of a Lady on Fire', country: 'France' },
  { id: 'wat', title: 'Woman at War', country: 'Iceland' },
];

const columns = [
  { key: 'title', label: 'Title' },
  { key: 'country', label: 'Country' },
];

const SomeOtherComponent = () => {
  return <div>Some other component</div>;
};

const Example = () => (
  <DataTableManagerProvider columns={columns}>
    <header>
      <DataTableManager />
      <SearchTextInput />
    </header>
    <main>
      <SomeOtherComponent />
      <DataTable rows={rows} />
    </main>
  </DataTableManagerProvider>
);

export default Example;
```

Reviewer: I would also add this note below the properties table:
```
> `*`: `DataTableManagerProvider` component accepts the same properties as the `DataTableManager`
```

Author: Updated here: https://github.com/commercetools/ui-kit/pull/2807/commits/e04a5226ed7a7bde3743ec5f6a9f1d607e92b562

Thank you

## Usage

There are two ways this component can be used:

> ### Attached

Reviewer: ```suggestion
### Atached
```

Author: This is the right spelling. Thanks! 🙃

Reviewer: Sorry

];

const Example = () => (

>   <DataTableManagerProvider>

Reviewer: I think we need to provide two different examples. The current one and another one with the detached behaviour where the `DataTable` and `DataTableManager` should be in different component's tree leaves.

Author: Updated here: https://github.com/commercetools/ui-kit/pull/2807/commits/bd8447871f0d99ac366b5adc9d865055c9a4506e

Thank you

export default Example;
```

## Using the DatatableManager when Detached from the DataTable

- To Detach the `<DatatableManager>` settings dropdown from the Datatable and position it anywhere else in the component tree, you would need to import a `DataTableManagerProvider` and wrap both components with the provider. This allows the settings component of the DatatableManager to be positioned anywhere within this provider.

```jsx
import DataTableManager, {
  DataTableManagerProvider,
} from '@commercetools-uikit/data-table-manager';
import DataTable from '@commercetools-uikit/data-table';

const rows = [
  { id: 'parasite', title: 'Parasite', country: 'South Korea' },
  { id: 'portrait', title: 'Portrait of a Lady on Fire', country: 'France' },
  { id: 'wat', title: 'Woman at War', country: 'Iceland' },
];

const columns = [
  { key: 'title', label: 'Title' },
  { key: 'country', label: 'Country' },
];

const SomeOtherComponent = () => {
  return <div>Some other component</div>;
};

const Example = () => (
  <DataTableManagerProvider columns={columns}>
    <DataTableManager />
    <SomeOtherComponent />
    <DataTable rows={rows} />

>   </DataTableManagerProvider>

Author: updated here: https://github.com/commercetools/ui-kit/pull/2807/commits/d1c167920c31da23e74275d1a3aa0912d539dc2d

Thank you

export default Example;
```

### Dettached

As mentioned earlier, the default behaviour is for the triggering element to be placed above the `<DataTable>` component (top-right corner), but there can be use cases where it would be needed to place that triggering element in a different place in the page. This is also possible but requires the usage of one more component (`DataTableManagerProvider`) in order to share the manager state between the manager panels and the `DataTable` component.

> 

Author: This makes sense. 
updated: https://github.com/commercetools/ui-kit/pull/2807/commits/e4aacf488d6e932f17cecedde608c092ac2e3c78

export default Example;
```

### Detached

As mentioned earlier, the default behaviour is for the triggering element to be placed above the `<DataTable>` component (top-right corner), but there can be use cases where it would be needed to place that triggering element in a different place in the page. This is also possible but requires the usage of one more component (`DataTableManagerProvider`) in order to share the manager state between the manager panels and the `DataTable` component.

> In this mode, you should pass the manager props to the `DataTableManagerProvider` component and the `DataTableManager` does not need to receive any prop; it can be placed anywhere in the component's tree without requiring any prop.

Reviewer: It would be great to include a breaking line between the two paragraphs.

export default Example;
```

### Detached

> As mentioned earlier, the default behaviour is for the triggering element to be placed above the `<DataTable>` component (top-right corner), but there can be use cases where it would be needed to place that triggering element in a different place in the page. This is also possible but requires the usage of one more component (`DataTableManagerProvider`) in order to share the manager state between the manager panels and the `DataTable` component.

Reviewer 2: nit: maybe this reads better:
```suggestion
As mentioned earlier, the default behaviour places the triggering element above the `<DataTable>` component (top-right corner), but there may be use cases where the triggering element needs to be positioned differently on the page. This is also possible but requires the usage of one more component (`DataTableManagerProvider`) in order to share the manager state between the manager panels and the `DataTable` component.
```

Author: updated here: https://github.com/commercetools/ui-kit/pull/2807/commits/37d6c7c20176cd563a6ad743408f2df3a22a752f

Thanks!
